src/common/components/repository-row/dropdowns/unregistered-dropdown.js

```diff
@@ -0,0 +1,171 @@
+  const authStoreFetchingDischarge = (
+    authStore.hasDischarge && !authStore.authenticated
+  );
+  const userMustSignAgreement = (
+    authStore.authenticated && authStore.signedAgreement === false
+  );
+
+  let caption;
+  if (registerNameStatus.success) {
+    caption = (
+      <div>
+        <TickIcon /> Registered successfully
+      </div>
+    );
+  } else if ( registerNameStatus.error
+    && registerNameStatus.error.json
+    && registerNameStatus.error.json.payload
+    && registerNameStatus.error.json.payload.code === 'already_registered') {
+    caption = (
+      <Message status='error'>
+        <p>Sorry, that name is already taken. Try a different name.</p>
+        <p className={ styles.helpText }>
+          If you think you should have sole rights to the name,
+          you can <a href={ FILE_NAME_CLAIM_URL } target='_blank'>file a claim</a>.
+        </p>
+      </Message>
+    );
+  } else if (registerNameStatus.error) {
+    caption = (
+      <Message status='error'>
+        { registerNameStatus.error.message }
+      </Message>
+    );
+  } else {
+    caption = (
+      <div>
+        To publish to the snap store, this repo needs a registered name.
+        { !authStore.authenticated &&
+          ' You need to sign in to Ubuntu One to register a name.'
+        }
+        { (authStoreFetchingDischarge || authStore.authenticated) &&
+          <div className={ styles.helpText }>
+            Lower-case letters, numbers, and hyphens only.
+          </div>
+        }
+        { userMustSignAgreement && <Agreement onChange={onSignAgreementChange}/> }
+      </div>
+    );
+  }
+
+  let actionDisabled;
+  let actionOnClick;
+  let actionSpinner = false;
+  let actionText;
+  if (authStore.isFetching) {
+    actionDisabled = true;
+    actionOnClick = () => {};
+    actionSpinner = true;
+    actionText = 'Checking...';
+  } else if (authStore.authenticated) {
+    actionDisabled = (
+      snapName === '' ||
+      registerNameStatus.isFetching ||
+      !!registerNameStatus.error
+    );
+    actionOnClick = onRegisterClick;
+    if (registerNameStatus.isFetching) {
+      actionSpinner = true;
+      actionText = 'Checking...';
+    } else {
+      actionText = 'Register name';
+    }
+  } else {
+    actionDisabled = !!registerNameStatus.error;
+    actionOnClick = onSignInClick.bind(this);
+    actionText = 'Sign in...';
+  }
+
+  return (
+    <Dropdown>
+      <Row>
+        <Data col="100">
+          { caption }
+        </Data>
+      </Row>
+      <Row>
+        <div className={ styles.buttonRow }>
+          <a onClick={onCancelClick} className={ styles.cancel }>
+            Cancel
+          </a>
+          <Button
+            appearance="positive"
+            disabled={actionDisabled}
+            onClick={actionOnClick}
+            isSpinner={actionSpinner}
+          >
+            { actionText }
+          </Button>
+        </div>
+      </Row>
+    </Dropdown>
+  );
+};
```

Comment:
This function could really benefit from being broken down into separate functions. It's still fine to keep it in the same file, I think.